pkg/services/ngalert/sender/sender.go

```diff
@@ -171,29 +176,104 @@ func buildNotifierConfig(cfg *ngmodels.AdminConfiguration) (*config.Config, erro
+
+	// Must sanitize labels in order otherwise resulting label set can be inconsistent when there are collisions.
+	for _, k := range sortedKeys(lbls) {
+		sanitizedLabelName, err := s.sanitizeLabelName(k)
+		if err != nil {
+			s.logger.Error("alert sending to external Alertmanager(s) contains an invalid label/annotation name that failed to sanitize, skipping", "name", k, "err", err)
+			continue
+		}
+
+		// There can be label name collisions after we sanitize. We check for this and attempt to make the name unique again using a short hash of the original name.
+		if _, ok := set[sanitizedLabelName]; ok {
+			sanitizedLabelName = sanitizedLabelName + fmt.Sprintf("_%.3x", md5.Sum([]byte(k)))
+			s.logger.Warn("alert contains duplicate label/annotation name after sanitization, appending unique suffix", "name", k, "new_name", sanitizedLabelName, "err", err)
+		}
 
-	// Prometheus does not allow spaces in labels or annotations while Grafana does, we need to make sure we
-	// remove them before sending the alerts.
-	for k, v := range alert.Alert.Labels {
-		ls = append(ls, labels.Label{Name: removeSpaces(k), Value: v})
+		set[sanitizedLabelName] = struct{}{}
+		ls = append(ls, labels.Label{Name: sanitizedLabelName, Value: lbls[k]})
 	}
 
-	for k, v := range alert.Annotations {
-		a = append(a, labels.Label{Name: removeSpaces(k), Value: v})
+	return ls
+}
+
+// sanitizeLabelName will fix a given label name so that it is compatible with prometheus alertmanager character restrictions.
+// Prometheus alertmanager requires labels to match ^[a-zA-Z_][a-zA-Z0-9_]*$.
+// Characters with an ASCII code < 127 will be replaced with an underscore (_), characters with ASCII code >= 127 will be replaced by their hex representation.
+// For backwards compatibility, whitespace will be removed instead of replaced with an underscore.
+func (s *ExternalAlertmanager) sanitizeLabelName(name string) (string, error) {
+	if len(name) == 0 {
+		return "", errors.New("label name cannot be empty")
 	}
 
-	return &notifier.Alert{
-		Labels:       ls,
-		Annotations:  a,
-		StartsAt:     time.Time(alert.StartsAt),
-		EndsAt:       time.Time(alert.EndsAt),
-		GeneratorURL: alert.Alert.GeneratorURL.String(),
+	if isValidLabelName(name) {
+		return name, nil
+	}
+
+	s.logger.Warn("alert sending to external Alertmanager(s) contains label/annotation name with invalid characters", "name", name)
+
+	// Remove spaces. We do this instead of replacing with underscore for backwards compatibility as this existed before the rest of this function.
+	sanitized := strings.Join(strings.Fields(name), "")
+
+	// Replace other invalid characters.
+	var buf strings.Builder
+	for i, b := range sanitized {
+		// From alertmanager LabelName.IsValid().
+		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
+			buf.WriteRune(b)
+		} else {
+			if b <= unicode.MaxASCII {
+				buf.WriteRune('_')
+			} else {
+				_, _ = fmt.Fprintf(&buf, "%x", b)
+			}
+		}
```

Comment:
I think we can reduce nesting a little bit, which will make it a little bit more readable
```suggestion
		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
			buf.WriteRune(b)
			continue
		} 
		if b <= unicode.MaxASCII {
			buf.WriteRune('_')
			continue
		} 
		_, _ = fmt.Fprintf(&buf, "%x", b)
```